### PR TITLE
nvmeadm: use the snafu crate to generate improved errors messages CAS…

### PR DESCRIPTION
**Pull request by blaisedias**
…-338

Replace use of the failure crate with snafu.
Fixes remaining part of https://github.com/openebs/Mayastor/issues/468

Add more context to errors generate, for example on IO errors,
add filename being accessed.

Change how connection in progress is handled by returning a specific error
so that the CSI driver can check explicitly, and use error kind for detection
instead of the magic number 114.

### CODE REVIEW

**Comment by blaisedias** [comment]
bors try

File: csi/src/error.rs [hunk 1]
```diff
@@ -86,3 +86,11 @@ impl From<FromUtf8Error> for DeviceError {
         }
     }
 }
+
+impl From<nvmeadm::error::NvmeError> for DeviceError {
```
Comment (tiagolobocastro) [inline csi/src/error.rs:90]: shouldn't these be done via Snafu now?

File: nvmeadm/src/lib.rs [hunk 2]
```diff
@@ -30,60 +28,37 @@ extern crate ioctl_gen;
...
         Ok(v) => Ok(v),
-        Err(_) => Err(std::io::Error::new(
-            ErrorKind::InvalidData,
-            format!(
-                "Failed to parse {}: {}",
-                path.as_path().to_str().unwrap(),
-                s
-            ),
-        )),
+        Err(_) => Err(NvmeError::ValueParseError {
```
Comment (tiagolobocastro) [inline nvmeadm/src/lib.rs:59]: might be useful to include the actual error in the Err

File: nvmeadm/tests/discovery_test.rs [hunk 3]
```diff
@@ -92,7 +91,7 @@ pub struct NvmfTarget {
 }
 
 impl NvmfTarget {
-    pub fn new(config_file: &str, nvmf_port: &str) -> Result<Self, Error> {
+    pub fn new(config_file: &str, nvmf_port: &str) -> Result<Self, ()> {
```
Comment (tiagolobocastro) [inline nvmeadm/tests/discovery_test.rs:94]: guess you could just return Self as it looks like this fn never fails (just panics)

File: csi/src/error.rs (hunk 1 above)
Comment (blaisedias) [reply csi/src/error.rs:90]: Yes, part of another change.

**Review by jonathan-teh — APPROVED** [review]

File: nvmeadm/tests/discovery_test.rs (hunk 3 above)
Comment (blaisedias) [reply nvmeadm/tests/discovery_test.rs]: Done.

File: nvmeadm/src/error.rs [hunk 4]
```diff
@@ -0,0 +1,51 @@
+use snafu::Snafu;
+
+#[derive(Debug, Snafu)]
+#[allow(missing_docs)]
+#[snafu(visibility = "pub(crate)")]
+pub enum NvmeError {
+    #[snafu(display("IO error:"))]
+    IoError { source: std::io::Error },
+    #[snafu(display("Failed to parse {} : {}", path, contents))]
```
[the remark this answers is not in the document]
Comment (blaisedias) [reply nvmeadm/src/error.rs]: Done

File: nvmeadm/src/error.rs [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+        filename: String,
+        source: std::io::Error,
+    },
+    #[snafu(display("nqn: {} not found", text))]
+    NqnNotFound { text: String },
+    #[snafu(display("No nvmf subsystems found"))]
+    NoSubsystems,
+    #[snafu(display("Connect in progress"))]
+    ConnectInProgress,
+    #[snafu(display("Nvme connect failed: {}, {}", filename, source))]
```
[the remark this answers is not in the document]
Comment (blaisedias) [reply nvmeadm/src/error.rs:24]: Done

File: nvmeadm/src/error.rs [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+#[allow(missing_docs)]
+#[snafu(visibility = "pub(crate)")]
+pub enum NvmeError {
+    #[snafu(display("IO error:"))]
+    IoError { source: std::io::Error },
+    #[snafu(display("Failed to parse {} : {}", path, contents))]
+    ValueParseError { path: String, contents: String },
+    #[snafu(display("Failed to parse value"))]
+    ParseError {},
+    #[snafu(display("File IO error:{}, {}", filename, source))]
```
[the remark this answers is not in the document]
Comment (blaisedias) [reply nvmeadm/src/error.rs]: Done

File: nvmeadm/src/error.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (blaisedias) [reply nvmeadm/src/error.rs]: Done

File: nvmeadm/src/lib.rs (hunk 2 above)
Comment (blaisedias) [reply nvmeadm/src/lib.rs]: Done

**Comment by blaisedias** [comment]
bors try

**Review by tiagolobocastro — APPROVED** [review]

**Comment by jonathan-teh** [comment]
CI is stuck:
```
2020-11-03T17:01:51.424724316+00:00 DEBUG nexus_create{name="rebuild_child_faulted" size=5242880 uuid=None children=["aio:///tmp/rebuild_child_faulted-disk0.img?blk_size=512", "aio:///tmp/rebuild_child_faulted-disk1.img?blk_size=512"]}: mayastor::bdev::nexus::nexus_bdev: close time.busy=11.1ms time.idle=31.7ms
2020-11-03T17:01:51.425193337+00:00 DEBUG mayastor::bdev::nexus::nexus_channel: rebuild_child_faulted: Creating IO channels at 0x562984fbc1e0```

**Comment by jonathan-teh** [comment]
bors try-

**Comment by blaisedias** [comment]
bors merge

**Comment by blaisedias** [comment]
bors merge

**Comment by blaisedias** [comment]
bors merge

**Pull request merged**